File: xscraper.py
import os
import sys
import json
import csv
import re
import pandas as pd
import logging
import gc
import socket
from datetime import datetime
from flask import Flask, request, jsonify

app = Flask(__name__)

# --- CONFIGURATION ---
BASE_DIR = os.path.dirname(os.path.abspath(__file__))
UPLOAD_PATH = os.path.join(BASE_DIR, "UPLOAD_FOLDER")

if not os.path.exists(UPLOAD_PATH):
    os.makedirs(UPLOAD_PATH)

# --- HEADERS ---
TWEET_HEADER = [
    'index_on_page', 'tweet_id', 'tweet_permalink_path', 'in_reply_to_user',
    'in_reply_to_tweet', 'quoted_tweet_id', 'user_screen_name',
    'user_id', 'user_name', 'user_location', 'user_timezone', 'user_lang',
    'user_bio', 'user_image_url', 'date_time', 'tweet_date',
    'coordinates_long', 'coordinates_lat', 'country', 'location_fullname',
    'location_name', 'location_type', 'raw_text', 'clear_text', 'user_verified',
    'hashtags', 'responses_to_tweeter', 'urls', 'user_mentions',
    'tweet_language', 'filter_level', 'is_retweet', 'is_quote', 'is_reply',
    'is_referenced', 'retweeted_tweet_id', 'retweeted_user_id',
    'retweeter_ids', 'is_message', 'has_image', 'media_link', 'has_video',
    'has_link', 'links', 'expanded_links', 'retweets', 'quotes', 'favorites',
    'replies', 'source', 'mentions_of_tweeter', 'context_annotations',
    'possibly_sensitive', 'conversation_id', 'withheld_copyright',
    'withheld_in_countries', 'withheld_scope', 'is_protected_or_deleted',
    'retweeter_api_cursor', 'views', 'blue_verified', 'video_views'
]

USER_HEADER = [
    'user_id', 'user_screen_name', 'user_name', 'user_lang', 'user_geo_enabled',
    'user_location', 'user_timezone', 'user_utc_offset', 'user_tweets', 
    'user_followers', 'user_following', 'user_friends', 'user_favorites', 
    'user_lists', 'user_bio', 'user_verified', 'user_protected', 
    'user_withheld_in_countries', 'user_withheld_scope', 'user_created', 
    'user_image_url', 'user_url', 'restricted_to_public', 'is_deleted', 
    'is_suspended', 'item_updated_time', 'not_in_search_results', 'blue_verified'
]

# --- UTILITIES ---
def initialize_record(record_type="tweet"):
    if record_type == "user":
        return {
            field: datetime.now().strftime('%Y-%m-%d %H:%M:%S') 
            if field == 'item_updated_time' 
            else None 
            for field in USER_HEADER
        }
    return {field: 0 if field in ['retweets', 'favorites', 'replies', 'quotes', 'views', 'video_views'] else None for field in TWEET_HEADER}

def format_x_date(date_str):
    if not date_str: return None
    try:
        dt = datetime.strptime(date_str, '%a %b %d %H:%M:%S %z %Y')
        return dt.strftime('%Y-%m-%d %H:%M:%S')
    except: return date_str

def clean_html(text):
    if not text: return ""
    return re.sub(r'<[^>]+>', '', text)

def safe_get(data, key):
    """Deep retrieval that handles nested dicts"""
    if not isinstance(data, dict): return None
    val = data.get(key)
    if isinstance(val, dict) and key in val:
        return val.get(key)
    return val

def clean_timezone(tz_str):
    """Ensure timezone is valid and not a date."""
    if not tz_str: return None
    tz_str = str(tz_str).strip()
    if re.search(r'20\d{2}', tz_str):
        return None
    return tz_str

# --- GLOBAL USER DATABASE & EXTRACTOR ---
users_db = {}

def extract_and_store_user(obj):
    if not isinstance(obj, dict): return
    if 'rest_id' not in obj: return
    
    legacy = obj.get('legacy') or {}
    core = obj.get('core') or {}
    avatar = obj.get('avatar') or {} 
    
    screen_name = legacy.get('screen_name') or core.get('screen_name')
    if not screen_name: return

    u_id = obj.get('rest_id')
    
    img = (
        safe_get(avatar, 'image_url') or 
        safe_get(legacy, 'profile_image_url_https') or 
        safe_get(obj, 'profile_image_url_https')
    )
    
    loc = safe_get(legacy, 'location') or safe_get(obj, 'location')
    
    created_raw = (
        legacy.get('created_at') or 
        core.get('created_at') or 
        obj.get('created_at')
    )
    
    raw_tz = legacy.get('time_zone')
    clean_tz = clean_timezone(raw_tz)

    # --- MAP VALUES ---
    # friends_count = The number of people this user is FOLLOWING
    # listed_count = The number of public lists this user is a member of
    following_count = legacy.get('friends_count') 
    lists_count = legacy.get('listed_count')

    if u_id not in users_db or (img and not users_db[u_id].get('user_image_url')):
        rec = initialize_record("user")
        rec.update({
            'user_id': str(u_id), # Ensure ID is string to prevent SciNotation
            'user_screen_name': screen_name,
            'user_name': legacy.get('name') or core.get('name'),
            'user_location': loc,
            'user_timezone': clean_tz,
            'user_image_url': img,
            'user_bio': legacy.get('description'),
            'user_followers': legacy.get('followers_count'),
            'user_following': following_count, 
            'user_friends': following_count,   
            'user_lists': lists_count,         
            'user_favorites': legacy.get('favourites_count'), 
            'user_tweets': legacy.get('statuses_count'),
            'user_verified': 1 if legacy.get('verified') else 0,
            'blue_verified': 1 if obj.get('is_blue_verified') else 0,
            'user_created': format_x_date(created_raw)
        })
        users_db[u_id] = rec

def recursive_signature_scan(data):
    if isinstance(data, dict):
        extract_and_store_user(data)
        for k, v in data.items():
            recursive_signature_scan(v)
    elif isinstance(data, list):
        for item in data:
            recursive_signature_scan(item)

# --- CORE PROCESSING LOGIC ---
def process_har_file(filename):
    if os.path.exists(filename):
        file_path = filename
    else:
        file_path = os.path.join(UPLOAD_PATH, filename)
    
    if not os.path.isfile(file_path):
        return {"error": f"File {filename} not found"}, 404

    try:
        print(f"Loading {filename}...")
        with open(file_path, 'r', encoding="utf-8") as f:
            content_j = json.loads(f.read())
        gc.collect() 
    except Exception as e:
        return {"error": f"Invalid JSON: {str(e)}"}, 400

    global users_db
    users_db = {}
    tweets = {}
    index_counter = 0

    print("Phase 1: Mining User Profiles via Signature Scan...")
    if 'log' in content_j and 'entries' in content_j['log']:
        for entry in content_j['log']['entries']:
            resp_text = entry.get('response', {}).get('content', {}).get('text')
            if not resp_text: continue
            try:
                data_json = json.loads(resp_text)
                recursive_signature_scan(data_json)
            except: continue

    print(f"Found {len(users_db)} unique users.")

    print("Phase 2: Processing Tweets...")
    if 'log' in content_j and 'entries' in content_j['log']:
        for entry in content_j['log']['entries']:
            url = entry.get('request', {}).get('url', '')
            if any(x in url for x in ['/graphql/', 'SearchTimeline', 'UserTweets']):
                try:
                    resp_text = entry.get('response', {}).get('content', {}).get('text')
                    if not resp_text: continue
                    data_json = json.loads(resp_text)
                    
                    data_root = data_json.get('data', {})
                    search_root = data_root.get('search_by_raw_query', {}).get('search_timeline', {}).get('timeline', {})
                    user_root = data_root.get('user', {}).get('result', {}).get('timeline_v2', {}).get('timeline', {})
                    conv_root = data_root.get('threaded_conversation_with_injections_v2', {})
                    
                    instructions = search_root.get('instructions', []) or \
                                   user_root.get('instructions', []) or \
                                   conv_root.get('instructions', [])

                    for ins in instructions:
                        if ins.get('type') == 'TimelineAddEntries':
                            for item in ins.get('entries', []):
                                eid = item.get('entryId', '')
                                if 'tweet-' not in eid and 'promoted' not in eid: continue
                                
                                content = item.get('content', {}).get('itemContent', {})
                                t_res = content.get('tweet_results', {}).get('result', {})
                                
                                if t_res.get('__typename') == 'TweetWithVisibilityResults':
                                    t_res = t_res.get('tweet', {})

                                if not t_res or 'legacy' not in t_res: continue
                                leg_t = t_res['legacy']
                                
                                u_res = t_res.get('core', {}).get('user_results', {}).get('result', {})
                                if u_res: extract_and_store_user(u_res)
                                
                                u_id = u_res.get('rest_id')
                                u_data = users_db.get(u_id, {})
                                
                                t_id = t_res.get('rest_id')
                                if t_id and t_id not in tweets:
                                    t_rec = initialize_record("tweet")
                                    
                                    # --- EXTRACT MENTIONS & REPLIES (FIX) ---
                                    ent = leg_t.get('entities', {})
                                    mentions_str = ""
                                    if 'user_mentions' in ent:
                                        # Join screen names with spaces: "user1 user2"
                                        mentions_str = " ".join([m['screen_name'] for m in ent['user_mentions']])
                                    
                                    reply_to_user = leg_t.get('in_reply_to_screen_name')
                                    reply_to_tweet = leg_t.get('in_reply_to_status_id_str')
                                    
                                    t_rec.update({
                                        'index_on_page': index_counter,
                                        'tweet_id': str(t_id), # String to force ID preservation
                                        'user_id': str(u_id),
                                        'user_screen_name': u_data.get('user_screen_name'),
                                        'user_name': u_data.get('user_name'),
                                        'user_location': u_data.get('user_location'),
                                        'user_timezone': u_data.get('user_timezone'),
                                        'user_image_url': u_data.get('user_image_url'),
                                        'raw_text': leg_t.get('full_text'),
                                        'clear_text': clean_html(leg_t.get('full_text')),
                                        'date_time': format_x_date(leg_t.get('created_at')),
                                        'retweets': leg_t.get('retweet_count', 0),
                                        'favorites': leg_t.get('favorite_count', 0),
                                        'replies': leg_t.get('reply_count', 0),
                                        'quotes': leg_t.get('quote_count', 0),
                                        'views': t_res.get('views', {}).get('count', 0),
                                        'source': clean_html(t_res.get('source', '')),
                                        'tweet_permalink_path': f"https://x.com/{u_data.get('user_screen_name')}/status/{t_id}",
                                        # Populated fields:
                                        'user_mentions': mentions_str,
                                        'in_reply_to_user': reply_to_user,
                                        'in_reply_to_tweet': str(reply_to_tweet) if reply_to_tweet else None,
                                        'is_reply': 1 if reply_to_tweet else 0,
                                        'is_retweet': 1 if leg_t.get('retweeted_status_result') else 0,
                                        'is_quote': 1 if leg_t.get('is_quote_status') else 0
                                    })
                                    
                                    if 'media' in ent:
                                        t_rec['media_link'] = " ".join([m['media_url_https'] for m in ent['media']])
                                        if any(m['type'] == 'photo' for m in ent['media']):
                                            t_rec['has_image'] = "1"
                                        if any(m['type'] == 'video' for m in ent['media']):
                                            t_rec['has_video'] = "1"
                                        v_views = 0
                                        for m in ent['media']:
                                            if 'mediaStats' in m and 'viewCount' in m['mediaStats']:
                                                v_views += m['mediaStats']['viewCount']
                                        t_rec['video_views'] = v_views

                                    tweets[t_id] = t_rec
                                    index_counter += 1
                except: continue

    del content_j
    gc.collect()

    base_name = os.path.splitext(os.path.basename(filename))[0]
    dt_postfix = datetime.now().strftime('%Y%m%d_%H%M%S')
    t_csv = f"{base_name}_tweets_{dt_postfix}.csv"
    u_csv = f"{base_name}_users_{dt_postfix}.csv"

    try:
        df_tweets = pd.DataFrame(list(tweets.values()))
        if df_tweets.empty: 
            df_tweets = pd.DataFrame(columns=TWEET_HEADER)
        else:
            # Ensure columns exist
            for col in TWEET_HEADER:
                if col not in df_tweets.columns: df_tweets[col] = None
            df_tweets = df_tweets[TWEET_HEADER]
            
        # Force string type for IDs to prevent Excel scientific notation conversion
        for col in ['tweet_id', 'user_id', 'in_reply_to_tweet', 'quoted_tweet_id']:
            if col in df_tweets.columns:
                df_tweets[col] = df_tweets[col].astype(str).replace('nan', '')

        df_tweets.to_csv(os.path.join(UPLOAD_PATH, t_csv), index=False, quoting=csv.QUOTE_ALL)

        df_users = pd.DataFrame(list(users_db.values()))
        if df_users.empty: 
            df_users = pd.DataFrame(columns=USER_HEADER)
        else:
             for col in USER_HEADER:
                if col not in df_users.columns: df_users[col] = None
             df_users = df_users[USER_HEADER]
             
        # Force string for User IDs
        if 'user_id' in df_users.columns:
             df_users['user_id'] = df_users['user_id'].astype(str).replace('nan', '')

        df_users.to_csv(os.path.join(UPLOAD_PATH, u_csv), index=False, quoting=csv.QUOTE_ALL)
    except Exception as e:
        return {"error": str(e)}, 500

    return {"message": "Success", "tweets_count": len(tweets), "files": [t_csv, u_csv]}

@app.route("/api/process", methods=["POST"])
def process_api():
    if 'file' not in request.files: return jsonify({"error": "No file"}), 400
    file = request.files['file']
    file_path = os.path.join(UPLOAD_PATH, file.filename)
    file.save(file_path)
    return jsonify(process_har_file(file.filename))

def find_free_port():
    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
        s.bind(('', 0))
        return s.getsockname()[1]

if __name__ == "__main__":
    if len(sys.argv) > 1:
        target_file = sys.argv[1]
        print(f"--- CLI Mode: Processing {target_file} ---")
        result = process_har_file(target_file)
        print(json.dumps(result, indent=2))
        print("Done.")
    else:
        print("--- Server Mode: Waiting for API requests ---")
        port = 5000
        try:
            app.run(debug=False, host='0.0.0.0', port=port)
        except OSError:
            port = find_free_port()
            print(f"Port 5000 busy. Starting on port {port}")
            app.run(debug=False, host='0.0.0.0', port=port)

File: xscraper.php
<?php
/**
 * xscraper.php - Comprehensive PHP Port
 * Fully replicates xscraper.py logic and data fields.
 */
ini_set('memory_limit', '1024M'); // Increased for large HAR files

class XScraperEngine
{
    private $uploadPath;
    private $usersDb = [];
    private $tweets = [];
    private $indexCounter = 0;

    private $tweetHeader = [
        'index_on_page', 'tweet_id', 'tweet_permalink_path', 'in_reply_to_user',
        'in_reply_to_tweet', 'quoted_tweet_id', 'user_screen_name',
        'user_id', 'user_name', 'user_location', 'user_timezone', 'user_lang',
        'user_bio', 'user_image_url', 'date_time', 'tweet_date',
        'coordinates_long', 'coordinates_lat', 'country', 'location_fullname',
        'location_name', 'location_type', 'raw_text', 'clear_text', 'user_verified',
        'hashtags', 'responses_to_tweeter', 'urls', 'user_mentions',
        'tweet_language', 'filter_level', 'is_retweet', 'is_quote', 'is_reply',
        'is_referenced', 'retweeted_tweet_id', 'retweeted_user_id',
        'retweeter_ids', 'is_message', 'has_image', 'media_link', 'has_video',
        'has_link', 'links', 'expanded_links', 'retweets', 'quotes', 'favorites',
        'replies', 'source', 'mentions_of_tweeter', 'context_annotations',
        'possibly_sensitive', 'conversation_id', 'withheld_copyright',
        'withheld_in_countries', 'withheld_scope', 'is_protected_or_deleted',
        'retweeter_api_cursor', 'views', 'blue_verified', 'video_views', 'user_geo_enabled'
    ];

    private $userHeader = [
        'user_id', 'user_screen_name', 'user_name', 'user_lang', 'user_geo_enabled',
        'user_location', 'user_timezone', 'user_utc_offset', 'user_tweets',
        'user_followers', 'user_following', 'user_friends', 'user_favorites',
        'user_lists', 'user_bio', 'user_verified', 'user_protected',
        'user_withheld_in_countries', 'user_withheld_scope', 'user_created',
        'user_image_url', 'user_url', 'restricted_to_public', 'is_deleted',
        'is_suspended', 'item_updated_time', 'not_in_search_results', 'blue_verified'
    ];

    public function __construct($uploadPath)
    {
        $this->uploadPath = rtrim($uploadPath, '/') . '/';
    }

    private function initializeRecord($type = "tweet")
    {
        $header = ($type === "user") ? $this->userHeader : $this->tweetHeader;
        $rec = array_fill_keys($header, "");

        if ($type === "user") {
            $rec['item_updated_time'] = date('Y-m-d H:i:s');
        }
        else {
            foreach (['retweets', 'favorites', 'replies', 'quotes', 'views', 'video_views'] as $key) {
                $rec[$key] = 0;
            }
        }
        return $rec;
    }

    private function formatXDate($dateStr, $onlyDate = false)
    {
        if (!$dateStr)
            return null;
        try {
            // Check if it's already in Y-m-d format
            if (preg_match('/^\d{4}-\d{2}-\d{2}$/', $dateStr)) {
                return $dateStr;
            }
            if (preg_match('/^\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2}$/', $dateStr)) {
                return $onlyDate ? explode(' ', $dateStr)[0] : $dateStr;
            }

            $date = DateTime::createFromFormat('D M d H:i:s O Y', $dateStr);
            if ($date) {
                return $onlyDate ? $date->format('Y-m-d') : $date->format('Y-m-d H:i:s');
            }

            // Fallback for other formats
            $date = new DateTime($dateStr);
            return $onlyDate ? $date->format('Y-m-d') : $date->format('Y-m-d H:i:s');
        }
        catch (Exception $e) {
            return $dateStr;
        }
    }

    private function cleanHtml($text)
    {
        return $text ? preg_replace('/<[^>]+>/', '', $text) : "";
    }

    private function extractAndStoreUser($obj)
    {
        if (!is_array($obj))
            return;
        if (isset($obj['user']) && is_array($obj['user']) && isset($obj['user']['rest_id'])) {
            $obj = $obj['user'];
        }
        if (!isset($obj['rest_id']))
            return;

        $legacy = $obj['legacy'] ?? [];
        $core = $obj['core'] ?? [];
        $relationship_counts = $obj['relationship_counts'] ?? [];
        $tweet_counts = $obj['tweet_counts'] ?? [];
        $action_counts = $obj['action_counts'] ?? [];
        $uId = (string)$obj['rest_id'];
        $screenName = $legacy['screen_name'] ?? ($core['screen_name'] ?? ($obj['screen_name'] ?? null));

        if (!$screenName)
            return;

        $img = $legacy['profile_image_url_https'] ?? ($obj['profile_image_url_https'] ?? ($obj['avatar']['image_url'] ?? null));

        if (!isset($this->usersDb[$uId])) {
            $this->usersDb[$uId] = $this->initializeRecord("user");
            $this->usersDb[$uId]['user_id'] = $uId;
        }

        $rec = & $this->usersDb[$uId];

        // Screen Name & Name
        if (!$rec['user_screen_name'])
            $rec['user_screen_name'] = $screenName;
        if (!$rec['user_name'])
            $rec['user_name'] = $legacy['name'] ?? ($core['name'] ?? ($obj['name'] ?? null));

        // Location & Bio
        $loc = $legacy['location'] ?? ($obj['location'] ?? ($core['location'] ?? null));
        if (is_array($loc) && isset($loc['location']))
            $loc = $loc['location'];
        if ($loc !== null && $loc !== "" && !is_array($loc)) {
            $rec['user_location'] = (string)$loc;
        }
        if (empty($rec['user_bio']))
            $rec['user_bio'] = $legacy['description'] ?? ($obj['profile_bio']['description'] ?? null);

        if (isset($legacy['protected'])) {
            $rec['user_protected'] = $legacy['protected'] ? 1 : 0;
        } elseif (isset($obj['privacy']['protected'])) {
            $rec['user_protected'] = $obj['privacy']['protected'] ? 1 : 0;
        }

        // Image
        if ($img && empty($rec['user_image_url']))
            $rec['user_image_url'] = $img;

        // Stats
        if (isset($legacy['followers_count'])) {
            $rec['user_followers'] = $legacy['followers_count'];
        } elseif (isset($relationship_counts['followers'])) {
            $rec['user_followers'] = $relationship_counts['followers'];
        }

        if (isset($legacy['friends_count'])) {
            $rec['user_following'] = $legacy['friends_count'];
            $rec['user_friends'] = $legacy['friends_count'];
        } elseif (isset($relationship_counts['following'])) {
            $rec['user_following'] = $relationship_counts['following'];
            $rec['user_friends'] = $relationship_counts['following'];
        }

        if (isset($legacy['listed_count'])) {
            $rec['user_lists'] = $legacy['listed_count'];
        } elseif (isset($relationship_counts['listed'])) {
            $rec['user_lists'] = $relationship_counts['listed'];
        } elseif (isset($tweet_counts['listed'])) {
            $rec['user_lists'] = $tweet_counts['listed'];
        }

        if (isset($legacy['favourites_count'])) {
            $rec['user_favorites'] = $legacy['favourites_count'];
        } elseif (isset($action_counts['favorites_count'])) {
            $rec['user_favorites'] = $action_counts['favorites_count'];
        }

        if (isset($legacy['statuses_count'])) {
            $rec['user_tweets'] = $legacy['statuses_count'];
        } elseif (isset($tweet_counts['tweets'])) {
            $rec['user_tweets'] = $tweet_counts['tweets'];
        }

        // Flags
        $verObj = $obj['verification'] ?? ($core['verification'] ?? []);
        $isVer = !empty($legacy['verified']) || !empty($verObj['verified']) || (!empty($verObj['verified_type']) && $verObj['verified_type'] !== 'Blue');
        if ($isVer)

            $rec['user_verified'] = 1;

        if (!empty($obj['is_blue_verified']) || !empty($core['is_blue_verified']))
            $rec['blue_verified'] = 1;

        if (!empty($legacy['geo_enabled']))
            $rec['user_geo_enabled'] = 1;

        // Date Created
        if (empty($rec['user_created'])) {
            $rec['user_created'] = $this->formatXDate($legacy['created_at'] ?? ($obj['created_at'] ?? ($core['created_at'] ?? null)), true);
        }

        // URL
        if (empty($rec['user_url'])) {
            $urlData = $legacy['entities']['url']['urls'][0] ?? ($obj['profile_bio']['entities']['url']['urls'][0] ?? null);
            $rec['user_url'] = $urlData['expanded_url'] ?? ($legacy['url'] ?? ($obj['website']['url'] ?? null));
        }
    }

    private function recursiveSignatureScan($data)
    {
        if (is_array($data)) {
            if (isset($data['rest_id']))
                $this->extractAndStoreUser($data);
            foreach ($data as $v)
                $this->recursiveSignatureScan($v);
        }
    }

    public function processHar($filePath)
    {
        $jsonRaw = file_get_contents($filePath);
        $content = json_decode($jsonRaw, true);
        unset($jsonRaw);

        if (!$content)
            return false;

        // Phase 1: Mine Users
        if (isset($content['log']['entries'])) {
            foreach ($content['log']['entries'] as $entry) {
                $resp = $entry['response']['content'] ?? null;
                $respText = $resp['text'] ?? null;
                if ($respText) {
                    if (($resp['encoding'] ?? '') === 'base64') {
                        $respText = base64_decode($respText);
                    }
                    $data = json_decode($respText, true);
                    if ($data)
                        $this->recursiveSignatureScan($data);
                }
            }
        }

        // Phase 2: Process Tweets
        if (isset($content['log']['entries'])) {
            foreach ($content['log']['entries'] as $entry) {
                $url = $entry['request']['url'] ?? '';
                if (preg_match('/(\/graphql\/|SearchTimeline|UserTweets)/', $url)) {
                    $resp = $entry['response']['content'] ?? [];
                    $respText = $resp['text'] ?? null;
                    if (!$respText)
                        continue;

                    if (($resp['encoding'] ?? '') === 'base64') {
                        $respText = base64_decode($respText);
                    }

                    $data = json_decode($respText, true);
                    if (!$data)
                        continue;

                    $instructions = $data['data']['search_by_raw_query']['search_timeline']['timeline']['instructions'] ??
                        ($data['data']['user']['result']['timeline_v2']['timeline']['instructions'] ??
                        ($data['data']['threaded_conversation_with_injections_v2']['instructions'] ?? []));

                    foreach ($instructions as $ins) {
                        $entries = $ins['entries'] ?? ($ins['moduleItems'] ?? []);
                        if (!$entries) continue;

                        foreach ($entries as $item) {
                            $itemsToProcess = [];
                            $eid = $item['entryId'] ?? ($item['item']['entryId'] ?? '');

                            if (strpos($eid, 'tweet-') !== false || strpos($eid, 'promoted') !== false) {
                                $itemsToProcess[] = $item;
                            } elseif (strpos($eid, 'conversationthread-') !== false || strpos($eid, 'module-') !== false) {
                                $subItems = $item['content']['items'] ?? ($item['item']['itemContent']['items'] ?? []);
                                foreach ($subItems as $sub) {
                                    if (isset($sub['item'])) {
                                        $itemsToProcess[] = $sub['item'];
                                    }
                                }
                            }

                                foreach ($itemsToProcess as $processItem) {
                                    $tRes = $processItem['content']['itemContent']['tweet_results']['result'] ?? 
                                            $processItem['itemContent']['tweet_results']['result'] ?? null;

                                    if (($tRes['__typename'] ?? '') === 'TweetWithVisibilityResults')
                                        $tRes = $tRes['tweet'] ?? null;

                                    if (!$tRes || !isset($tRes['legacy']))
                                        continue;

                                    $leg = $tRes['legacy'];
                                    $tId = (string)$tRes['rest_id'];

                                if (isset($tRes['core']['user_results']['result'])) {
                                    $this->extractAndStoreUser($tRes['core']['user_results']['result']);
                                }

                                if ($tId && !isset($this->tweets[$tId])) {
                                    $uId = (string)($tRes['core']['user_results']['result']['rest_id'] ?? '');
                                    $uData = $this->usersDb[$uId] ?? [];

                                    $fullDate = $this->formatXDate($leg['created_at'] ?? null);
                                    $rawText = $leg['full_text'] ?? '';

                                    // Extraction logic
                                    preg_match_all('/#(\w+)/', $rawText, $hMatches);
                                    $entities = $leg['entities'] ?? [];
                                    $linksList = [];
                                    $expandedList = [];
                                    if (isset($entities['urls'])) {
                                        foreach ($entities['urls'] as $u) {
                                            if (!empty($u['url']))
                                                $linksList[] = $u['url'];
                                            if (!empty($u['expanded_url']))
                                                $expandedList[] = $u['expanded_url'];
                                        }
                                    }

                                    $mentionsList = [];
                                    if (isset($entities['user_mentions'])) {
                                        foreach ($entities['user_mentions'] as $m) {
                                            if (!empty($m['screen_name']))
                                                $mentionsList[] = $m['screen_name'];
                                        }
                                    }

                                    $mediaSource = $leg['extended_entities'] ?? ($leg['entities'] ?? []);
                                    $mediaLinks = [];
                                    $hasImg = null;
                                    $hasVid = null;
                                    $vViews = 0;
                                    if (isset($mediaSource['media'])) {
                                        foreach ($mediaSource['media'] as $m) {
                                            $mediaLinks[] = $m['media_url_https'] ?? '';
                                            if (($m['type'] ?? '') === 'photo')
                                                $hasImg = "1";
                                            if (in_array($m['type'] ?? '', ['video', 'animated_gif']))
                                                $hasVid = "1";
                                            $vViews += $m['mediaStats']['viewCount'] ?? 0;
                                        }
                                    }

                                    $rec = $this->initializeRecord("tweet");
                                    $rec = array_merge($rec, [
                                        'index_on_page' => $this->indexCounter++,
                                        'tweet_id' => $tId,
                                        'user_id' => $uId,
                                        'user_screen_name' => $uData['user_screen_name'] ?? null,
                                        'user_name' => $uData['user_name'] ?? null,
                                        'user_location' => $uData['user_location'] ?? null,
                                        'user_image_url' => $uData['user_image_url'] ?? null,
                                        'user_bio' => $uData['user_bio'] ?? null,
                                        'user_verified' => $uData['user_verified'] ?? 0,
                                        'blue_verified' => $uData['blue_verified'] ?? 0,
                                        'user_geo_enabled' => $uData['user_geo_enabled'] ?? 0,
                                        'raw_text' => $rawText,
                                        'clear_text' => $this->cleanHtml($rawText),
                                        'date_time' => $fullDate,
                                        'tweet_date' => $fullDate ? explode(' ', $fullDate)[0] . " 00:00:00" : null,
                                        'tweet_language' => $leg['lang'] ?? null,
                                        'in_reply_to_tweet' => $leg['in_reply_to_status_id_str'] ?? null,
                                        'in_reply_to_user' => $leg['in_reply_to_screen_name'] ?? null,
                                        'is_reply' => (!empty($leg['in_reply_to_status_id_str'])) ? "1" : null,
                                        'is_retweet' => (!empty($leg['retweeted_status_id_str'])) ? "1" : null,
                                        'retweeted_tweet_id' => $leg['retweeted_status_id_str'] ?? null,
                                        'is_quote' => (!empty($leg['is_quote_status']) || !empty($leg['quoted_status_id_str'])) ? "1" : null,
                                        'quoted_tweet_id' => $leg['quoted_status_id_str'] ?? null,
                                        'hashtags' => !empty($hMatches[1]) ? implode(' ', $hMatches[1]) : null,
                                        'coordinates_lat' => $leg['geo']['coordinates'][0] ?? null,
                                        'coordinates_long' => $leg['geo']['coordinates'][1] ?? null,
                                        'country' => $leg['place']['country'] ?? null,
                                        'location_fullname' => $leg['place']['full_name'] ?? null,
                                        'location_name' => $leg['place']['name'] ?? null,
                                        'location_type' => $leg['place']['place_type'] ?? null,
                                        'has_link' => (!empty($linksList)) ? "1" : null,
                                        'links' => implode(' ', $linksList),
                                        'expanded_links' => implode(' ', $expandedList),
                                        'user_mentions' => implode(' ', $mentionsList),
                                        'retweets' => $leg['retweet_count'] ?? 0,
                                        'favorites' => $leg['favorite_count'] ?? 0,
                                        'replies' => $leg['reply_count'] ?? 0,
                                        'quotes' => $leg['quote_count'] ?? 0,
                                        'views' => $tRes['views']['count'] ?? 0,
                                        'source' => $this->cleanHtml($tRes['source'] ?? ''),
                                        'media_link' => implode(' ', $mediaLinks),
                                        'has_image' => $hasImg,
                                        'has_video' => $hasVid,
                                        'video_views' => $vViews > 0 ? $vViews : 0,
                                        'tweet_permalink_path' => "https://x.com/" . ($uData['user_screen_name'] ?? 'i') . "/status/$tId"
                                    ]);

                                    if (!empty($leg['geo']) || !empty($leg['place'])) {
                                        if (isset($this->usersDb[$uId])) {
                                            $this->usersDb[$uId]['user_geo_enabled'] = 1;
                                        }
                                        if (empty($uData['user_geo_enabled'])) {
                                            $rec['user_geo_enabled'] = 1;
                                        }
                                    }

                                    $this->tweets[$tId] = $rec;
                                }
                                }
                            }
                        }
                    }
                }
            }

        $base = pathinfo($filePath, PATHINFO_FILENAME);
        $postfix = date('Ymd_His');
        $tFile = "{$base}_tweets_{$postfix}.csv";
        $uFile = "{$base}_users_{$postfix}.csv";
        $jFile = "{$base}_data_{$postfix}.json";

        $this->writeCsv($this->uploadPath . $tFile, $this->tweetHeader, $this->tweets);
        $this->writeCsv($this->uploadPath . $uFile, $this->userHeader, $this->usersDb);
        file_put_contents($this->uploadPath . $jFile, json_encode([
            'tweets' => array_values($this->tweets),
            'users' => array_values($this->usersDb)
        ], JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES));

        return [
            'tweets' => $tFile,
            'users' => $uFile,
            'json' => $jFile,
            'tweet_count' => count($this->tweets),
            'user_count' => count($this->usersDb)
        ];
    }

    private function writeCsv($path, $headers, $data)
    {
        $fp = fopen($path, 'w');
        fputcsv($fp, $headers, ",", "\"", "\\");
        foreach ($data as $row) {
            $line = [];
            foreach ($headers as $h) {
                $line[] = $row[$h] ?? '';
            }
            fputcsv($fp, $line, ",", "\"", "\\");
        }
        fclose($fp);
    }
}